### DISCUSSION TITLE
start/stop non-OpenGL thread from the viewer

### DISCUSSION THREAD

**Discussion by xarthurx** [libigl/libigl#1753] [Q&A] [answered]
I've read multiple issues relating to threading for non-OpenGL process but am still not very clear if imgui can do this or not.
So sorry for the spam here.

**Need**:
In my UI, I have two buttons: `startServer` and `stopServer`, which will start/stop a WebSocket Server.
The Server will receive data from time to time and send things back to the client side after processing the data.

The main UI can be blocked or not blocked after starting the Server, but I would like to have the functionality to press `stopServer`.
Meanwhile, after I press the `startServer`, the main UI is frozen after the WebSockets started.

From what I understand, I may need to put it in another thread.
But does `imgui` or `libigl` allows this?

![image](https://user-images.githubusercontent.com/1921878/111185018-d7227f80-85b1-11eb-8447-bb487f2b81f6.png)

**Comment by jdumas** [comment, marked as answer]
I'm not sure what you expect us to answer. You need to spawn and manage the threads yourself if you want the ImGui buttons to be non-blocking. The only thing you need to be careful is to not call any OpenGL function from another thread than the main OpenGL thread.
